backend/vapid.js

```diff
@@ -0,0 +1,104 @@
+import sbp from '@sbp/sbp'
+
+let vapidPublicKey: string
+let vapidPrivateKey: Object
+
+// The Voluntary Application Server Identification (VAPID) email field is "a
+// stable identity for the application server" that "can be used by a push
+// service to establish behavioral expectations for an application server"
+// RFC 8292
+if (!process.env.VAPID_EMAIL) {
+  console.warn('Missing VAPID identification. Please set VAPID_EMAIL to a value like "mailto:some@example".')
+}
+const vapid = { VAPID_EMAIL: process.env.VAPID_EMAIL || 'mailto:test@example.com' }
+
+export const initVapid = async () => {
+  const vapidKeyPair = await sbp('chelonia/db/get', '_private_immutable_vapid_key').then(async (vapidKeyPair: string): Promise<[Object, string]> => {
+    if (!vapidKeyPair) {
+      // Generate a new ECDSA key pair
+      const keyPair = await crypto.subtle.generateKey(
+        {
+          name: 'ECDSA',
+          namedCurve: 'P-256' // Use P-256 curve
+        },
+        true, // Whether the key is extractable
+        ['sign', 'verify'] // Usages
+      )
+
+      // Export the private key
+      const serializedKeyPair = await Promise.all([
+        crypto.subtle.exportKey('jwk', keyPair.privateKey),
+        crypto.subtle.exportKey('raw', keyPair.publicKey).then((key) =>
+          // $FlowFixMe[incompatible-call]
+          Buffer.from(key).toString('base64url')
+        )
+      ])
+
+      return sbp('chelonia/db/set', '_private_immutable_vapid_key', JSON.stringify(serializedKeyPair)).then(() => {
+        return [keyPair.privateKey, serializedKeyPair[1]]
+      })
+    }
+
+    const serializedKeyPair = JSON.parse(vapidKeyPair)
+    return [
+      await crypto.subtle.importKey(
+        'jwk',
+        serializedKeyPair[0],
+        { name: 'ECDSA', namedCurve: 'P-256' },
+        false,
+        ['sign']
+      ),
+      serializedKeyPair[1]
+    ]
+  })
+
+  vapidPrivateKey = vapidKeyPair[0]
+  vapidPublicKey = vapidKeyPair[1]
+}
+
+const generateJwt = async (endpoint: URL): Promise<string> => {
+  const now = Date.now() / 1e3 | 0
+
+  const audience = endpoint.origin
```

Comment:
Claude:

> According to the Web Push Protocol spec (RFC 8292), the audience claim should be the push service's URL origin. While using endpoint.origin works for most push services, some might require the exact scheme + host + port. It would be safer to ensure the audience matches the exact requirements of the push service.

At first I thought this was nonsense, but then I looked at what `endpoint` is [supposed to be](https://devdocs.io/dom/pushsubscription/endpoint), and [the docs](https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription/endpoint) say it's a string, not a `URL`. Yet here it's being used as a `URL`, and maybe with these potential issues too, so am leaving this comment.

Reply:
This is nonsense. Using `origin` will give the information needed for URLs.

> say it's a string, not a URL

Correct, but the subscription info is wrapped and we convert it to a URL for convenience.

Reply:
Where does this happen? I see it comes from `server.clients` => `client.endpoint`, but I don't see where `server.clients` is filled...

Reply:
That happens in `subscriptionInfoWrapper`.

Reply:
Could you then add a link to the JWT spec where it says that `audience` needs to be the `origin`?

Reply:
Yes, but that's not the JWT spec. That'd be the VAPID spec.